docs/proposals/007-variadic-generics.md

```diff
@@ -0,0 +1,662 @@
+{
+    void process(int data) { ... }
+}
+
+void user(int myData[N])
+{
+    // Define an instance of the functor, and pass it to `process`.
+    MyProcessorFunc functor = ...;
+    process(functor, myData);
+}
+```
+
+While this can work, it requires a lot of boilterplate from the user. For each shape of callback, the user must define
+a separate interface. We can reduce this boilterplate if the system has builtin support for `IFunc`:
+
+```
+// The callback function `p` is represented as a functor conforming to the `IProcessor` interface.
+void process<TProcessor:IFunc<void, int>>(TProcessor p, int data[N])
+{
+    for (int i = 0; i < N; i++)
+        p.process(data[i]);
+}
+
+// Define the functor as a type that conforms to `IProcessor`.
+struct MyProcessorFunc : IFunc<void, int>
+{
+    void process(int data) { ... }
+}
+
+void user(int myData[N])
+{
+    // Define an instance of the functor, and pass it to `process`.
+    MyProcessorFunc functor = ...;
+    process(functor, myData);
+}
+```
+
+The above code eliminates the user defined interface by using the builtin `IFunc` interface. By making `IFunc` builtin,
+we can open the path for the compiler to synthesize conformances to `IFunc` for ordinary functions and in the future
+add support for lambda expressions that automatically conform to `IFunc`, further simplify the user code into something like:
+
+```
+// The callback function `p` is represented as a functor conforming to the `IProcessor` interface.
+void process<TProcessor:IFunc<void, int>>(TProcessor p, int data[N])
+{
+    for (int i = 0; i < N; i++)
+        p.process(data[i]);
+}
+
+void user(int myData[N])
+{
+    process((int x)=>{...}, myData);
+}
+```
+
+Related Work
+------------
+
+Variadic generics is an advance type system feature that is missinng in many modern languages including C# and Rust.
+Swift adds support for variadic generics recently in late 2022/2023, and this proposal largely follows Swift's design.
+C++ has variadic templates that achieves similar results within the template system.
+Rust supports variadics in a macro system layered above its core type system. While this can solve many user issues,
+we decided to not go through this path because macros and templates must be expanded before core type checking, which means that
+they can't integrate nicely in modules and compiled into IR independently of their use sites.
+
+
+Proposed Approach
+-----------------
+
+Slang can follow Swift's solution for variadic generics. A user can define a variadic generic with the syntax:
+
+```
+void myFunc<each T>(expand each T v) {...}
+```
+
+The code above defines a generic function that has a __generic type pack parameter__ `T` with the `each` keyword before `T`.
+The function's parameter list is defined as `expand each T v`, which should be interpreted as a parameter `v` whose type is
+`expand each T`. `expand each T` is a type that represents a pack of types. A parameter whose type is a pack of types can
```

Comment:
Now I know what `expand each T` is.
Can you please also explain or give a definition of `each T`?